File: tests/cpp/test_interpol.cpp
/**
 * 
 */

#include "mole.h"

int main(void)
{

    u16 k;
    u32 m, n, o;

    ivec pdc1 = {0, 0};
    ivec nc1 = {0, 0};
    ivec ndc1 = {1, 1};

    ivec pdc2 = {0, 0, 0, 0};
    ivec nc2 = {0, 0, 0, 0};
    ivec ndc2 = {1, 1, 1, 1};

    ivec pdc3 = {0, 0, 0, 0, 0, 0};
    ivec nc3 = {0, 0, 0, 0, 0, 0};
    ivec ndc3 = {1, 1, 1, 1, 1, 1};

    for (k = 2; k < 3; k += 2)
    {
        m = 2 * k + 1;
        n = m + 1;
        o = n + 1;

        /****************
        1-D Interpolators
        ****************/

        // Periodic
        InterpolCtoF picf1(k, m, pdc1, nc1);
        InterpolCtoN picn1(k, m, pdc1, nc1);
        InterpolFtoC pifc1(k, m, pdc1, nc1);
        InterpolNtoC pinc1(k, m, pdc1, nc1);

        picf1.save("picf1D.csv", arma::csv_ascii);
        picn1.save("picn1D.csv", arma::csv_ascii);
        pifc1.save("pifc1D.csv", arma::csv_ascii);
        pinc1.save("pinc1D.csv", arma::csv_ascii);

        // Nonperiodic
        InterpolCtoF nicf1(k, m, ndc1, nc1);
        InterpolCtoN nicn1(k, m, ndc1, nc1);
        InterpolFtoC nifc1(k, m, ndc1, nc1);
        InterpolNtoC ninc1(k, m, ndc1, nc1);

        nicf1.save("nicf1D.csv", arma::csv_ascii);
        nicn1.save("nicn1D.csv", arma::csv_ascii);
        nifc1.save("nifc1D.csv", arma::csv_ascii);
        ninc1.save("ninc1D.csv", arma::csv_ascii);

        /****************
        2-D Interpolators
        ****************/

        // Periodic
        InterpolCtoF picf2(k, m, n, pdc2, nc2);
        InterpolCtoN picn2(k, m, n, pdc2, nc2);
        InterpolFtoC pifc2(k, m, n, pdc2, nc2);
        InterpolNtoC pinc2(k, m, n, pdc2, nc2);

        picf2.save("picf2D.csv", arma::csv_ascii);
        picn2.save("picn2D.csv", arma::csv_ascii);
        pifc2.save("pifc2D.csv", arma::csv_ascii);
        pinc2.save("pinc2D.csv", arma::csv_ascii);

        // Nonperiodic
        InterpolCtoF nicf2(k, m, n, ndc2, nc2);
        InterpolCtoN nicn2(k, m, n, ndc2, nc2);
        InterpolFtoC nifc2(k, m, n, ndc2, nc2);
        InterpolNtoC ninc2(k, m, n, ndc2, nc2);

        nicf2.save("nicf2D.csv", arma::csv_ascii);
        nicn2.save("nicn2D.csv", arma::csv_ascii);
        nifc2.save("nifc2D.csv", arma::csv_ascii);
        ninc2.save("ninc2D.csv", arma::csv_ascii);

        /****************
        3-D Interpolators
        ****************/

        // Periodic
        InterpolCtoF picf3(k, m, n, o, pdc3, nc3);
        InterpolCtoN picn3(k, m, n, o, pdc3, nc3);
        InterpolFtoC pifc3(k, m, n, o, pdc3, nc3);
        InterpolNtoC pinc3(k, m, n, o, pdc3, nc3);

        picf3.save("picf3D.csv", arma::csv_ascii);
        picn3.save("picn3D.csv", arma::csv_ascii);
        pifc3.save("pifc3D.csv", arma::csv_ascii);
        pinc3.save("pinc3D.csv", arma::csv_ascii);

        // Nonperiodic
        InterpolCtoF nicf3(k, m, n, o, ndc3, nc3);
        InterpolCtoN nicn3(k, m, n, o, ndc3, nc3);
        InterpolFtoC nifc3(k, m, n, o, ndc3, nc3);
        InterpolNtoC ninc3(k, m, n, o, ndc3, nc3);

        nicf3.save("nicf3D.csv", arma::csv_ascii);
        nicn3.save("nicn3D.csv", arma::csv_ascii);
        nifc3.save("nifc3D.csv", arma::csv_ascii);
        ninc3.save("ninc3D.csv", arma::csv_ascii);
    }

    return 0;
}

File: tests/cpp/test_interpol.m
clear; clc; close all;
addpath("../../src/matlab_octave");

tol = 1e-10;

for k = 2:2:3

    m = 2 * k + 1;
    n = m + 1;
    o = n + 1;

    icfcpp = readmatrix("picf1D.csv");
    icncpp = readmatrix("picn1D.csv");
    ifccpp = readmatrix("pifc1D.csv");
    inccpp = readmatrix("pinc1D.csv");

    if ~all(icfcpp - interpolCentersToFacesD1DPeriodic(k, m) < tol, "all")
        disp("1-D, k = " + k + ", Periodic Centers to Faces: Pass")
    else
        disp("1-D, k = " + k + ", Periodic Centers to Faces: Fail")
    end

    if ~all(icncpp - interpolCentersToNodes1DPeriodic(k, m) < tol, "all")
        disp("1-D, k = " + k + ", Periodic Centers to Nodes: Pass")
    else
        disp("1-D, k = " + k + ", Periodic Centers to Nodes: Fail")
    end

    if ~all(ifccpp - interpolFacesToCentersG1DPeriodic(k, m) < tol, "all")
        disp("1-D, k = " + k + ", Periodic Faces to Centers: Pass")
    else
        disp("1-D, k = " + k + ", Periodic Faces to Centers: Fail")
    end

    if ~all(inccpp - interpolNodesToCenters1DPeriodic(k, m) < tol, "all")
        disp("1-D, k = " + k + ", Periodic Nodes to Centers: Pass")
    else
        disp("1-D, k = " + k + ", Periodic Nodes to Centers: Fail")
    end

    icfcpp = readmatrix("picf2D.csv");
    icncpp = readmatrix("picn2D.csv");
    ifccpp = readmatrix("pifc2D.csv");
    inccpp = readmatrix("pinc2D.csv");

    if all(abs(icfcpp - interpolCentersToFacesD2DPeriodic(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Periodic Centers to Faces: Pass")
    else
        disp("2-D, k = " + k + ", Periodic Centers to Faces: Fail")
    end

    if all(abs(icncpp - interpolCentersToNodes2DPeriodic(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Periodic Centers to Nodes: Pass")
    else
        disp("2-D, k = " + k + ", Periodic Centers to Nodes: Fail")
    end

    if all(abs(ifccpp - interpolFacesToCentersG2DPeriodic(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Periodic Faces to Centers: Pass")
    else
        disp("2-D, k = " + k + ", Periodic Faces to Centers: Fail")
    end

    if all(abs(inccpp - interpolNodesToCenters2DPeriodic(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Periodic Nodes to Centers: Pass")
    else
        disp("2-D, k = " + k + ", Periodic Nodes to Centers: Fail")
    end

    icfcpp = readmatrix("picf3D.csv");
    icncpp = readmatrix("picn3D.csv");
    ifccpp = readmatrix("pifc3D.csv");
    inccpp = readmatrix("pinc3D.csv");

    if all(abs(icfcpp - interpolCentersToFacesD3DPeriodic(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Periodic Centers to Faces: Pass")
    else
        disp("3-D, k = " + k + ", Periodic Centers to Faces: Fail")
    end

    if all(abs(icncpp - interpolCentersToNodes3DPeriodic(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Periodic Centers to Nodes: Pass")
    else
        disp("3-D, k = " + k + ", Periodic Centers to Nodes: Fail")
    end

    if all(abs(ifccpp - interpolFacesToCentersG3DPeriodic(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Periodic Faces to Centers: Pass")
    else
        disp("3-D, k = " + k + ", Periodic Faces to Centers: Fail")
    end

    if all(abs(inccpp - interpolNodesToCenters3DPeriodic(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Periodic Nodes to Centers: Pass")
    else
        disp("3-D, k = " + k + ", Periodic Nodes to Centers: Fail")
    end



    icfcpp = readmatrix("nicf1D.csv");
    icncpp = readmatrix("nicn1D.csv");
    ifccpp = readmatrix("nifc1D.csv");
    inccpp = readmatrix("ninc1D.csv");

    if all(abs(icfcpp - interpolCentersToFacesD1D(k, m)) < tol, "all")
        disp("1-D, k = " + k + ", Nonperiodic Centers to Faces: Pass")
    else
        disp("1-D, k = " + k + ", Nonperiodic Centers to Faces: Fail")
    end

    if all(abs(icncpp - interpolCentersToNodes1D(k, m)) < tol, "all")
        disp("1-D, k = " + k + ", Nonperiodic Centers to Nodes: Pass")
    else
        disp("1-D, k = " + k + ", Nonperiodic Centers to Nodes: Fail")
    end

    if all(abs(ifccpp - interpolFacesToCentersG1D(k, m)) < tol, "all")
        disp("1-D, k = " + k + ", Nonperiodic Faces to Centers: Pass")
    else
        disp("1-D, k = " + k + ", Nonperiodic Faces to Centers: Fail")
    end

    if all(abs(inccpp - interpolNodesToCenters1D(k, m)) < tol, "all")
        disp("1-D, k = " + k + ", Nonperiodic Nodes to Centers: Pass")
    else
        disp("1-D, k = " + k + ", Nonperiodic Nodes to Centers: Fail")
    end

    icfcpp = readmatrix("nicf2D.csv");
    icncpp = readmatrix("nicn2D.csv");
    ifccpp = readmatrix("nifc2D.csv");
    inccpp = readmatrix("ninc2D.csv");

    if all(abs(icfcpp - interpolCentersToFacesD2D(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Nonperiodic Centers to Faces: Pass")
    else
        disp("2-D, k = " + k + ", Nonperiodic Centers to Faces: Fail")
    end

    if all(abs(icncpp - interpolCentersToNodes2D(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Nonperiodic Centers to Nodes: Pass")
    else
        disp("2-D, k = " + k + ", Nonperiodic Centers to Nodes: Fail")
    end

    if all(abs(ifccpp - interpolFacesToCentersG2D(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Nonperiodic Faces to Centers: Pass")
    else
        disp("2-D, k = " + k + ", Nonperiodic Faces to Centers: Fail")
    end

    if all(abs(inccpp - interpolNodesToCenters2D(k, m, n)) < tol, "all")
        disp("2-D, k = " + k + ", Nonperiodic Nodes to Centers: Pass")
    else
        disp("2-D, k = " + k + ", Nonperiodic Nodes to Centers: Fail")
    end

    icfcpp = readmatrix("nicf3D.csv");
    icncpp = readmatrix("nicn3D.csv");
    ifccpp = readmatrix("nifc3D.csv");
    inccpp = readmatrix("ninc3D.csv");

    if all(abs(icfcpp - interpolCentersToFacesD3D(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Nonperiodic Centers to Faces: Pass")
    else
        disp("3-D, k = " + k + ", Nonperiodic Centers to Faces: Fail")
    end

    if all(abs(icncpp - interpolCentersToNodes3D(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Nonperiodic Centers to Nodes: Pass")
    else
        disp("3-D, k = " + k + ", Nonperiodic Centers to Nodes: Fail")
    end

    if all(abs(ifccpp - interpolFacesToCentersG3D(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Nonperiodic Faces to Centers: Pass")
    else
        disp("3-D, k = " + k + ", Nonperiodic Faces to Centers: Fail")
    end

    if all(abs(inccpp - interpolNodesToCenters3D(k, m, n, o)) < tol, "all")
        disp("3-D, k = " + k + ", Nonperiodic Nodes to Centers: Pass")
    else
        disp("3-D, k = " + k + ", Nonperiodic Nodes to Centers: Fail")
    end

end